src/main/java/com/google/sps/servlets/NewsServlet.java

```diff
@@ -10,42 +10,135 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import com.google.sps.data.Article;
+import java.net.URL;
+import java.net.HttpURLConnection;
+import java.util.Scanner;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonElement;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.net.URLEncoder;
+import java.util.HashMap;
 
 /**
  * Servlet for generating news articles from keywords.
  */
 @WebServlet("/news")
 public class NewsServlet extends HttpServlet {
-  private List<String> keywords = new ArrayList<>();
-  
-  // Hardcoded list of articles. 
-  // In future implementation, will be generated based on key words.
-  private List<Article> articles = new ArrayList<>(
-      List.of(
-        new Article.Builder(
-            "Bolton says Trump turned a blind eye to the coronavirus pandemic",
-            "https://www.cnn.com/2020/06/24/politics/john-bolton-interview-cnntv/index.html")
-          .withWriter("Zachary Cohen")
-          .withDate("06/24/2020")
-          .build(),
-        new Article.Builder(
-            "Hearing goes off the rails when lawmaker keeps banging table",
-            "https://www.cnn.com/videos/politics/2020/06/24/louie-gohmert-bangs-table-judiciary-hearing-vpx.cnn")
-          .withDate("06/24/2020")
-          .build()
-      )   
+
+  private String API_KEY = "test";
+  private int numArticlesPerKeyword = 3;
+  private List<String> keywords = new ArrayList<String>(
+    List.of("Black Lives Matter", "COVID-19")
   );
+  private HashMap<String, List<Article>> articleMap = new HashMap<String, List<Article>>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = makeNewsJson(articles);
+    for (String keyword : keywords) {
+      keyword = encodeTerm(keyword);
+      addArticlesForTerm(keyword);
+    }
+    String json = makeNewsJson();
     response.getWriter().println(json);
   }
 
-  private String makeNewsJson(List<Article> articleList) {
+ /**
+  * Encodes the given term so that it can fit in a url.
+  * @return An encoded string
+  */
+  private String encodeTerm(String term) {
+    try {
+      return URLEncoder.encode(term, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  private String makeNewsJson() {
     Gson gson = new Gson();
-    String json = gson.toJson(articleList);
-    json = String.format("{\"articles\": %s}", json);
+    String json = gson.toJson(articleMap);
+    json = "{\"articles\": " + json + "}";  
```

Comment:
[Optional] Instead of using string formatting, you could make it a map before passing it to gson:
```suggestion
    Map<String, Object> articles = Map.of("articles", articleMap);
    String json = gson.toJson(articles);
```

Reply:
I just realized that Map.of is only Java 9+, so you can't use it. But still I would use the Map.

Reply:
It seemed to be working for me.